### ISSUE TITLE
Custom models produce nonsensical (constant) output

### ISSUE THREAD

**Issue by lschmelzeisen** [guillaume-be/rust-bert#228]
I'm currently trying to use the [distilbert-base-german-cased](https://huggingface.co/distilbert-base-german-cased) model from the hugging face models hub with rust-bert for `[MASK]`-filling. I'm hoping to do this via [`DistilBertModelMaskedLM`](https://docs.rs/rust-bert/latest/rust_bert/distilbert/struct.DistilBertModelMaskedLM.html).

For this, I have produced the following minimal example implementation, which is adapted from the [`distilbert_masked_lm`](https://github.com/guillaume-be/rust-bert/blob/23c5d9112abbde66109ebebf6113e81144e89d8c/tests/distilbert.rs#L43) test:

```rust
use itertools::zip;
use itertools::Itertools;
use rust_bert::distilbert::{DistilBertConfig, DistilBertModelMaskedLM};
use rust_bert::resources::{LocalResource, Resource};
use rust_bert::Config;
use rust_tokenizers::tokenizer::{BertTokenizer, MultiThreadedTokenizer, TruncationStrategy};
use rust_tokenizers::vocab::Vocab;
use std::path::PathBuf;
use tch::{nn, no_grad, Device, Tensor};

extern crate anyhow;

fn main() -> anyhow::Result<()> {
    // Resources paths
    let model_name = "distilbert-base-german-cased";
    let config_resource = Resource::Local(LocalResource {
        local_path: PathBuf::from(format!("data/{model_name}/config.json")),
    });
    let vocab_resource = Resource::Local(LocalResource {
        local_path: PathBuf::from(format!("data/{model_name}/vocab.txt")),
    });
    let weights_resource = Resource::Local(LocalResource {
        local_path: PathBuf::from(format!("data/{model_name}/rust_model.ot")),
    });

    // Set-up masked LM model
    let device = Device::cuda_if_available();
    let mut vs = nn::VarStore::new(device);
    let tokenizer: BertTokenizer = BertTokenizer::from_file(
        vocab_resource.get_local_path()?.to_str().unwrap(),
        true,
        true,
    )?;
    let config = DistilBertConfig::from_file(config_resource.get_local_path()?);
    let distil_bert_model = DistilBertModelMaskedLM::new(&vs.root(), &config);
    vs.load(weights_resource.get_local_path()?)?;

    // Define input
    let input = [
        "Looks like one [MASK] is missing",
        "It\'s like comparing [MASK] to apples",
        "Es scheint eine [MASK] zu fehlen",
        "Das ist, als würde man [MASK] mit Äpfeln vergleichen",
    ];
    let tokenized_input = tokenizer.encode_list(&input, 128, &TruncationStrategy::LongestFirst, 0);
    let max_len = tokenized_input
        .iter()
        .map(|input| input.token_ids.len())
        .max()
        .unwrap();
    let tokenized_input = tokenized_input
        .iter()
        .map(|input| input.token_ids.clone())
        .map(|mut input| {
            input.extend(vec![0; max_len - input.len()]);
            Tensor::of_slice(&input)
        })
        .collect::<Vec<_>>();
    let input_tensor = Tensor::stack(tokenized_input.as_slice(), 0).to(device);

    // Forward pass
    let model_output = no_grad(|| {
        distil_bert_model
            .forward_t(Some(&input_tensor), None, None, false)
            .unwrap()
    });

    // Print results
    let mask_token = tokenizer.vocab().token_to_id("[MASK]");
    for i in 0..input.len() {
        let mask_pos = tokenized_input[i]
            .iter::<i64>()?
            .position(|token| token == mask_token)
            .unwrap();
        println!(
            "\nInput: {}",
            tokenized_input[i]
                .iter::<i64>()?
                .map(|token| tokenizer.vocab().id_to_token(&token))
                .join(" ")
        );
        let (scores, indices) = model_output
            .prediction_scores
            .get(i as i64)
            .get(mask_pos as i64)
            .topk(5, 0, true, true);
        for (score, index) in zip(scores.iter::<f64>()?, indices.iter::<i64>()?) {
            println!("  {} ({:.4})", tokenizer.vocab().id_to_token(&index), score);
        }
    }

    Ok(())
}
```

(The last two input examples in `input` are the German translation of the two English examples.)

For `distilbert-base-german-cased`, I receive the following output:

```plain
Input: [CLS] look ##s like one [MASK] is miss ##ing [SEP] [PAD] [PAD] [PAD]
  und (8.8536)
  - (8.5275)
  , (8.3060)
  . (7.4291)
  ##e (7.4181)

Input: [CLS] it ' s like comp ##aring [MASK] to app ##les [SEP] [PAD]
  und (8.8727)
  - (8.5572)
  , (8.1767)
  . (7.4396)
  ##e (7.3744)

Input: [CLS] es scheint eine [MASK] zu fehlen [SEP] [PAD] [PAD] [PAD] [PAD] [PAD]
  und (8.8790)
  - (8.6022)
  , (8.2564)
  ##e (7.4573)
  gab (7.3988)

Input: [CLS] das ist , als wurde man [MASK] mit ap ##feln vergleichen [SEP]
  und (8.9918)
  - (8.6382)
  , (8.3216)
  . (7.6484)
  ##e (7.5349)
```

Clearly, the predicted words are not actually dependent on the surrounding context and the predictions are almost constant irregardless of what the input sentence is.

If `model_name` is changed to `distilbert-base-cased`, everything works as expected:

```plain
Input: [CLS] looks like one [MASK] is missing [SEP] [PAD] [PAD] [PAD] [PAD] [PAD] [PAD] [PAD] [PAD] [PAD] [PAD] [PAD] [PAD] [PAD] [PAD] [PAD]
  tooth (7.2761)
  syllable (7.0870)
  word (6.6179)
  thing (6.4401)
  item (6.3338)

Input: [CLS] it ' s like comparing [MASK] to apples [SEP] [PAD] [PAD] [PAD] [PAD] [PAD] [PAD] [PAD] [PAD] [PAD] [PAD] [PAD] [PAD] [PAD]
  apples (12.0375)
  fruits (7.0120)
  monkeys (6.9805)
  them (6.6027)
  grapes (6.5141)

Input: [CLS] es s ##che ##int e ##ine [MASK] zu f ##eh ##len [SEP] [PAD] [PAD] [PAD] [PAD] [PAD] [PAD] [PAD] [PAD] [PAD] [PAD]
  und (6.7265)
  ##ß (6.2130)
  ##ben (6.0917)
  gut (5.7937)
  , (5.2297)

Input: [CLS] das is ##t , al ##s w ##ur ##de man [MASK] mit a ##pf ##el ##n ve ##rg ##le ##iche ##n [SEP]
  ##nen (8.7707)
  ##chester (7.9587)
  ##d (7.9419)
  ##gen (7.9234)
  ##gel (7.9127)
```

(Naturally, the predictions for the two German examples are not good, but that is because the model is not trained for that.)

Here the files `config.json` and `vocab.txt` in the directories `data/distilbert-base-cased` and `data/distilbert-base-german-cased` are downloaded directly from the model hub.
The `rust_model.ot` files are generated from downloaded `pytorch_model.bin` files using the [`convert_model.py`](https://github.com/guillaume-be/rust-bert/blob/master/utils/convert_model.py) script, which I think is the intended purpose of that script.

Is there anything wrong with my usage of the conversion script or the rust-bert crate? Or is this something the crate is not capable of doing? What would be necessary to implement it?

For reference, here are the completions of `distilbert-base-german-cased` from the HuggingFace-hosted Fill-MASK example, which is what I would also like rust-bert's output to be.

![Screenshot from 2022-02-23 12-01-24](https://user-images.githubusercontent.com/488697/155320927-f04e7c78-bc99-441e-b5cd-1292e623a410.png)
![Screenshot from 2022-02-23 12-01-41](https://user-images.githubusercontent.com/488697/155320933-8de8bce8-c1e8-4c7a-82bf-e8d0b4184bd0.png)

**Comment by guillaume-be** [comment]
Thank you @lschmelzeisen for raising this issue and the very detailed report - this was very useful and I could reproduce the error.
This is due to a bug with models using sinusoidal embeddings which would then ignore the weights in the state dictionary. For _distilbert-base-uncased_, the `sinusoidal_pos_embds` field is set to `false`: https://huggingface.co/distilbert-base-uncased/blob/main/config.json#L18, which means the position embeddings get loaded without issue. On the other hand, the German version of Distilbert has this variable set to `true`: https://huggingface.co/distilbert-base-german-cased/blob/main/config.json#L19.

The Rust library generates sinusoidal embeddings and ignores the finetuned version of these embeddings when loading the model weights. 

I am pushing an update (https://github.com/guillaume-be/rust-bert/pull/230) that would load all the weights from the state dictionary (including position embeddings when they are used for initialization) in which should solve the issue. In the meantime, you can also solve your issue in the current crate version by manually turning this flag off in the config before building the model:

```rust
fn main() -> anyhow::Result<()> {
    // Resources paths
    [...]
    let mut config = DistilBertConfig::from_file(config_resource.get_local_path()?);
    config.sinusoidal_pos_embds = false;
    let distil_bert_model = DistilBertModelMaskedLM::new(&vs.root(), &config);
    vs.load(weights_resource.get_local_path()?)?;
    [...]
```

**Comment by lschmelzeisen** [comment]
Thanks for the swift and in-depth answer and fix!

I can confirm that using `config.sinusoidal_pos_embds = false;` produces good results. They are sensible, but naturally still a bit different from what I got via the web interface. But I suspect that your update fixes this (I don't have a way to test that right now).

As far as I'm concerned this issue can be closed now. But maybe you want to wait until a new version is released.

**Comment by guillaume-be** [comment]
I am getting the same results as the web interface. To get comparable results you would have to:
- Turn off lowercasing and accent stripping:
```rust
    let tokenizer: BertTokenizer = BertTokenizer::from_file(
        vocab_resource.get_local_path()?.to_str().unwrap(),
        false,
        false,
    )?;
```
- Send unpadded input as this can impact the results. Sending one input at a time (e.g. `let input = ["Es scheint eine [MASK] zu fehlen"]`) gives me:
```
  Spur (7.5854)
  Antwort (7.5204)
  Lösung (7.4258)
  Erklärung (7.0258)
  Alternative (6.4393)
```

**Comment by lschmelzeisen** [comment]
Oh, good catch. Can confirm. Thanks again!
